src/main/kotlin/org/rust/lang/core/types/ty/TyPointer.kt

```diff
@@ -10,7 +10,10 @@ import org.rust.lang.core.resolve.ImplLookup
 data class TyPointer(val referenced: Ty, val mutability: Mutability) : Ty {
 
     override fun unifyWith(other: Ty, lookup: ImplLookup): UnifyResult =
-        if (other is TyPointer) referenced.unifyWith(other.referenced, lookup) else UnifyResult.fail
+        if (other is TyPointer && mutability == other.mutability)
```

Comment:
This fails unification when the mutability does not match.
I am not fully sure about this.
I tried resolving fuzzy in this case but I still get no good resolving because of "Multiple implementations"

Reply:
I am currently not sure how to use the exact/fuzzy match with resolving but I will look into what i can figure out.

Reply:
Hmm, I was mistaken here. Really this code does not compile:
```rust
struct S;
impl<'a> Tr for &'a S {}
trait Tr { fn foo(&self) {} }

fn main() {
    let a = S;
    let b = &mut a;
    b.foo();
}
```
But it will compile if we replace `&mut a` to just `&a`

Reply:
And try the same with TyReference. It may fix #1667

Reply:
I am not sure I understand what this means. Is the TyPointer unification change OK as it is now?

Yes adapting TyReference the same ways as TyPointer fixes the linked issue :-)
In the past I have tried the same code in TyReference but could not find a case where it mattered so i did not include it in my patch.

I will rebase for the associated type bugfix and push tomorrow when I have more time.

Reply:
You understood correctly, it is OK =)